sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -74,8 +81,24 @@ class ExpressionSet protected(
   }
 
   override def -(elem: Expression): ExpressionSet = {
-    val newBaseSet = baseSet.clone().filterNot(_ == elem.canonicalized)
-    val newOriginals = originals.clone().filterNot(_.canonicalized == elem.canonicalized)
+    if (elem.deterministic) {
+      val newBaseSet = baseSet.clone().filterNot(_ == elem.canonicalized)
+      val newOriginals = originals.clone().filterNot(_.canonicalized == elem.canonicalized)
+      new ExpressionSet(newBaseSet, newOriginals)
+    } else {
+      new ExpressionSet(baseSet.clone(), originals.clone())
+    }
+  }
+
+  override def --(elems: GenTraversableOnce[Expression]): ExpressionSet = {
```

Comment:
Could you explain why we should override `--`? I still don't quite get that.

Reply:
me too, it seems only a minor optimization, to avoid cloning the `baseSet` and `originals` many times when removing multiple non-deterministic expressions. I think it's a premature optimization and let's remove it for now.

Reply:
No need to add this in this PR. Please remove it.